### ISSUE TITLE
openstack plugin fails to remove the VM it launched when startup timeout fails

### ISSUE THREAD

**Issue by dbilling** [jenkinsci/openstack-cloud-plugin#310]
The openstack plugin has a startup timeout option that purposefully abandons newly launched agent VMs after if a productive agent connection is not achieved before the timeout hits.  However, if this timeout triggers for whatever reason, I've noticed the rejected/timed out nodes hang around, consuming cloud resources (and costing money) until I manually notice them and deleted them.    Best I can tell, there is no current option to have this plugin delete the VM that it launched and subsequently abandoned due to the timeout.

Before I build some custom automation around detecting and terminating these hung nodes outside of the plugin, I thought  it would be useful to report this issue here on the chance it could be fixed for all plugin users -- if the plugin launched a node and then fails to use it, shouldn't the plugin delete the abandoned VM (either by default or by optional parameter)?

**Comment by olivergondza** [comment]
@dbilling, that is expected to be correctly deleted. Check Jenkins server logs for the Server name and id and attach it here so this can be investigated on when exactly it timed out. Also, what does [Cloud Stats](https://plugins.jenkins.io/cloud-stats/) says about those servers?

**Comment by dbilling** [comment]
@olivergondza,  I can recreate this a number of different ways - the easiest are either by interrupting ssh connectivity or by removing access to something jenkins needs (e.g. java) .    Jenkins continues reconnect attempts until the startup timeout.  After the timeout, you'd expect the openstack plugin would remove the node.  But it doesn't, the node hangs forever.

after the timeout, cloud stats says "This agent is off line because Jenkins failed to launch the agent process on it".    
the log for the agent node is all typical startup retry stuff and has a final line that says "Launcher failed to bring the node online within timeout." 

this state persists forever, the node is now hung in an unproductive state, manual intervention is needed to delete the node.

**Comment by olivergondza** [comment]
Hmm, these should be detected every ten minutes here: https://github.com/jenkinsci/openstack-cloud-plugin/blob/129f931c3e4bf20a678f02c063d47db8db4dcc4e/plugin/src/main/java/jenkins/plugins/openstack/compute/JCloudsCleanupThread.java#L99. Do you see the message logged by your Jenkins? Any other interesting logs by `JCloudsCleanupThread`?

If you run this in groovy console (https://plugins.jenkins.io/openstack-cloud/#troubleshooting), does it find affected computers?
```groovy
jenkins.plugins.openstack.compute.JCloudsComputer.getAll().collect { it.name + ": " + it.idle}
```

What version of a plugin you are using? Did this used to work correctly before?

**Comment by dbilling** [comment]
the plugin version is: 2.56   

I don't know if this ever worked correctly for me in the past or not -- I only recently ran into this when I had a agent node with an issue and wondered why the bad node hung around forever.
 
I've recreated today by simply offering an agent node that doesn't have java installed to get the info you've asked for...

When I run the following script from the script console, it does find the "bad", affected node,SCCC-Jenkins-Node-3445.
```
jenkins.plugins.openstack.compute.JCloudsComputer.getAll().collect { it.name + ":  idle:" + it.idle + "  offline cause:" + it.offlineCause}
[SCCC-Jenkins-Node-3445:  idle:true  offline cause:This agent is offline because Jenkins failed to launch the agent process on it.]
```

Every 10 mins. when cleanup thread to run, all I see for  jenkins.plugins.openstack.compute.JCloudsCleanupThread is:
```
Started OpenStack slave cleanup
Finished OpenStack slave cleanup. 11,173 ms
```
and nothing more.

**Comment by dbilling** [comment]
I think the problem has to do with offline cause vs. fatal offline cause.
```
jenkins.plugins.openstack.compute.JCloudsComputer.getAll().collect { it.name + ":  idle:" + it.idle + "  offline cause:" + it.getOfflineCause() + " fatal offline cause:" + it.getFatalOfflineCause()}

Result: [SCCC-Jenkins-Node-3445:  idle:true  offline cause:This agent is offline because Jenkins failed to launch the agent process on it. fatal offline cause:null]
```

**Comment by olivergondza** [comment]
@dbilling, Spot on!

We need to look into this deeper and either make this offline cause a fatal one, or understand why the node does not get into other fatal offline cause after the timeout is over. I suspect the node have this offline cause for all repeated launch attempts even _before_ the timeout is over, so declaring this cause a fatal one will not work well.

What is the type in your case?

```groovy
jenkins.plugins.openstack.compute.JCloudsComputer.getAll().collect { it.name + "  offline cause:" + it.getOfflineCause().getClass()}
```

**Comment by dbilling** [comment]
Hi @olivergondza,  

here's the output:
Result: [SCCC-Jenkins-Node-102  offline cause:class hudson.slaves.OfflineCause$LaunchFailed]

**Comment by dbilling** [comment]
@olivergondza  anything else needed here to debug? Any possibility of a fix in the next plugin version?

**Comment by olivergondza** [comment]
The information here seems complete to reproduce in unittest and fix. Thanks for that. However, not sure when I will have cycles to have a look at this.

**Comment by jikortus** [comment]
Hi @olivergondza,
do you have any idea when this could be fixed, please? I've been hitting this misbehaviour as well and it would be really helpful to have this fixed (for example this time the available IP addresses were exhausted, causing the VMs to hang, and I needed to investigate what actually happened, instead of having the slaves reprovisioned again after some time). Thank you!

**Comment by olivergondza** [comment]
Released in version 2.62. Soon in your UC!

Thank you all for reporting this.

**Comment by jikortus** [comment]
Thank you very much, @olivergondza!
